pkg/storage/cacher.go

```diff
@@ -402,10 +403,20 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 			glog.Errorf("invalid object for filter: %v", obj)
 			return false
 		}
-		if !strings.HasPrefix(objKey, key) {
-			return false
+
+		segments := strings.Split(objKey, "/")
```

Comment:
>  since key "ns1/pod1" is also prefix of objKey "ns1/pod11"+"/"

"ns1/pod1" -> "ns1/pods1/"
"ns1/pod11" -> "ns1/pod11/"

Those are impl. details. Parsing through dirs is fine with me though.
The fundamental problem is that filterFunction doesn't have any tests.